kauri/kauri.go

```diff
@@ -0,0 +1,320 @@
+
+// NoFaultTree is a tree without any faulty replicas in the tree, which is currently supported.
+// TreeWithFaults is a tree with faulty replicas, not yet implemented.
+const (
+	NoFaultTree    TreeType = 1
+	TreeWithFaults TreeType = 2
+)
+
+// Kauri structure contains the modules for kauri protocol implementation.
+type Kauri struct {
+	configuration          *backend.Config
+	server                 *backend.Server
+	blockChain             modules.BlockChain
+	crypto                 modules.Crypto
+	eventLoop              *eventloop.EventLoop
+	logger                 logging.Logger
+	opts                   *modules.Options
+	synchronizer           modules.Synchronizer
+	leaderRotation         modules.LeaderRotation
+	tree                   TreeConfiguration
+	initDone               bool
+	aggregatedContribution hotstuff.QuorumSignature
+	blockHash              hotstuff.Hash
+	currentView            hotstuff.View
+	senders                []hotstuff.ID
+	nodes                  map[hotstuff.ID]*kauripb.Node
+	isAggregationSent      bool
+}
+
+// New initializes the kauri structure
+func New() modules.Kauri {
+	return &Kauri{nodes: make(map[hotstuff.ID]*kauripb.Node)}
+}
+
+// InitModule initializes the Handel module.
+func (k *Kauri) InitModule(mods *modules.Core) {
+	mods.Get(
+		&k.configuration,
+		&k.server,
+		&k.blockChain,
+		&k.crypto,
+		&k.eventLoop,
+		&k.logger,
+		&k.opts,
+		&k.leaderRotation,
+		&k.synchronizer,
+	)
+	k.opts.SetShouldUseKauri()
+	k.eventLoop.RegisterObserver(backend.ConnectedEvent{}, func(_ any) {
+		k.postInit()
+	})
+	k.eventLoop.RegisterHandler(ContributionRecvEvent{}, func(event any) {
+		k.OnContributionRecv(event.(ContributionRecvEvent))
+	})
+}
+
+func (k *Kauri) postInit() {
+	k.logger.Info("Handel: Initializing")
+	kauripb.RegisterKauriServer(k.server.GetGorumsServer(), serviceImpl{k})
+	k.initializeConfiguration()
+}
+
+func (k *Kauri) initializeConfiguration() {
+	kauriCfg := kauripb.ConfigurationFromRaw(k.configuration.GetRawConfiguration(), nil)
+	for _, n := range kauriCfg.Nodes() {
+		k.nodes[hotstuff.ID(n.ID())] = n
+	}
+	k.tree = CreateTree(k.configuration.Len(), k.opts.ID())
+
+	// pIDs := make(map[hotstuff.ID]int)
+	// for id := range k.configuration.ActiveReplicas() {
+	// 	pIDs[id] = index
+	// 	index++
+	// }
+	idMappings := make(map[hotstuff.ID]int)
+	for i := 0; i < k.configuration.Len(); i++ {
+		idMappings[hotstuff.ID(i+1)] = i
+	}
+	k.tree.InitializeWithPIDs(idMappings)
+	k.initDone = true
+	k.senders = make([]hotstuff.ID, 0)
+}
+
+// Begin starts dissemination of proposal and aggregation of votes.
+func (k *Kauri) Begin(pc hotstuff.PartialCert, p hotstuff.ProposeMsg) {
+	if !k.initDone {
+		k.eventLoop.DelayUntil(backend.ConnectedEvent{}, func() { k.Begin(pc, p) })
+		return
+	}
+	k.reset()
+	k.blockHash = pc.BlockHash()
+	k.currentView = p.Block.View()
+	k.aggregatedContribution = pc.Signature()
+	ids := k.randomizeIDS(k.blockHash, k.leaderRotation.GetLeader(k.currentView))
```

Comment:
This can be moved to the tree and merged with `InitializeWithPIDs()` as outlined in the [comment below](https://github.com/relab/hotstuff/pull/89/files#r1135014800).